firewood/src/merkle.rs

```diff
@@ -1175,6 +1180,180 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
         self.store.flush_dirty()
     }
+
+    pub fn get_iter<K: AsRef<[u8]>>(
+        &self,
+        key: Option<K>,
+        root: DiskAddress,
+    ) -> Result<MerkleKeyValueStream<'_, K, S>, MerkleError> {
+        // TODO: if DiskAddress::is_null() ...
+        Ok(MerkleKeyValueStream {
+            starting_key: key,
+            merkle_root: root,
+            merkle: self,
+            current_node: Default::default(),
+            parents: Default::default(),
+            current_key: Default::default(),
+        })
+    }
+}
+
+pub struct MerkleKeyValueStream<'a, K, S> {
+    starting_key: Option<K>,
+    merkle_root: DiskAddress,
+    merkle: &'a Merkle<S>,
+    // current node is the node that was last returned from poll_next
+    current_node: Option<ObjRef<'a>>,
+    // parents hold pointers up the tree to the parents
+    parents: Vec<(ObjRef<'a>, u8)>,
+    // this is the last key returned
+    // TODO: improve this; we can probably just adjust current_key as we walk up
+    // and down the merkle tree rather than fully rebuilding it each time
+    current_key: Option<Vec<u8>>,
+}
+
+impl<'a, K: AsRef<[u8]> + Unpin, S: shale::ShaleStore<node::Node> + Send + Sync> Stream
+    for MerkleKeyValueStream<'a, K, S>
+{
+    type Item = Result<(Vec<u8>, Vec<u8>), api::Error>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let mut pinned_self = pin!(self);
+        if let Some(key) = pinned_self.starting_key.take() {
+            // skip to this key, which must exist
+            // TODO: support finding the next key after K
+            let root_node = pinned_self
+                .merkle
+                .get_node(pinned_self.merkle_root)
+                .map_err(|e| api::Error::InternalError(e.into()))?;
+
+            (pinned_self.current_node, pinned_self.parents) = pinned_self
+                .merkle
+                .get_node_and_parents_by_key(root_node, &key)
+                .map_err(|e| api::Error::InternalError(e.into()))?;
+            if pinned_self.current_node.as_ref().is_none() {
+                return Poll::Ready(None);
+            }
+            pinned_self.current_key = Some(key.as_ref().to_vec());
+            return match pinned_self.current_node.as_ref().unwrap().inner() {
+                NodeType::Branch(branch) => Poll::Ready(Some(Ok((
+                    key.as_ref().to_vec(),
+                    branch.value.to_owned().unwrap().to_vec(),
+                )))),
+
+                NodeType::Leaf(leaf) => {
+                    Poll::Ready(Some(Ok((key.as_ref().to_vec(), leaf.1.to_vec()))))
+                }
+                NodeType::Extension(_) => todo!(),
+            };
+        }
+        // The current node might be none if the tree is empty or we happen to be at the end
+        let current_node = match pinned_self.current_node.take() {
+            None => return Poll::Ready(None),
```

Comment:
I can't do the empty key test yet because it's not implemented. I need to add code that walks down the leftmost branch every time and returns that key, which doesn't currently exist.

I'm thinking about not using the existing get_node_and_parents_by_key or refactoring it so it supports:
- Option<K> so you can get the lowest leaf node
- returning the next highest key if the key can't be found (maybe as an option? or wrapping in another function to discard the key if it isn't found?)